DEVELOPMENT-GUIDE.md

```diff
@@ -90,4 +90,71 @@ While a `jni_borrow` exists, ACCP may be in a critial region and thus any use of
+What's more, it is easy to not notice that there is a Java exception pending.
+So long as you do not interact with `JNIEnv`/`raii_env` directly, you shouldn't need to worry about Java exceptions as the ACCP objects/methods which use them already have appropriate checks which throw C++ exceptions (specifically `java_ex`) instead.
+If you need to throw an exception, it should almost alwasy be a C++ exception and thrown using the `throw_java_ex` or `throw_openssl` methods. (The former is more efficient but the latter MUST be used when the exception is due to an error state reported by OpenSSL.)
+
+Openssl has its *own* separate error handling in the form of a thread-local queue.
+This has caused bugs in the past where consuming code has not noticed that errors were present on the stack and so later calls incorrectly saw old and irrelevant Openssl errors.
+`throw_openssl` correctly checks *and clears* the OpenSSL error queue prior to throwing the C++ exception.
+It also will use the OpenSSL provided error message if available.
+If there is no OpenSSL error or applicable message, it will use the provided default message.
+
+Making sure that the OpenSSL error queue is empty prior to returning from native code is critical.
+This is why *during coverage tests* we enable extra assertions which will terminate the process if any unhandled OpenSSL errors are present.
+These assertions are only present during the coverage tests as they can be expensive (especially on multi-threaded systems).
+
+### Critical Regions
+There are two different ways to read data from Java arrays and Strings: Copying, or Critical Regions.
+* [String Critical methods](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetStringCritical_ReleaseStringCritical)
+* [Array Critical methods](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetPrimitiveArrayCritical_ReleasePrimitiveArrayCritical)
+
+ACCP generally uses critical regions because they are (usually) faster.
+When you are in a critical region the JVM (generally) pins the object being handled to a specific memory location and gives you direct access to the underlying data.
+This means that the JVM is unable to do many of its memory management functions (as they require moving objects) and is operating essentially in a degraded mode. (No garbage collection, etc.)
+So, when you are in a critical region there are exceedingly few operations you can safely do.
+Essentially, all you can do is methods which manipulate your critical region until you get out of it.
+It is **extremely important** that you do not allocate *any* Java memory (such as creating Java objects, like a Java Exception), call *any* Java methods (which you shouldn't be doing anyway), or take any actions which may block.
+It is also **very important** that you don't spend too much time in a critical region.
```

Comment:
"too much time" - maybe you can be a little more specific. Don't spend so much time that you risk delaying GC while other application threads are amassing garbage that needs to be cleaned. Typically it is a good idea to stay below a millisecond of call duration.

Reply:
Sounds like a good rule of thumb. I somehow doubt ACCP always achieves that right now (as crypto can be expensive), but that's a good goal for us to aim for.